### ISSUE TITLE
useDebounce or throttle?

### ISSUE THREAD

**Issue by bebbi** [xnimorz/use-debounce#22]
Am I wrong thinking `useDebounce` works more like `useThrottle`?

On first call, it immediately sets the value to the target value without debouncing.

I was trying to use it for a value which, when not falsy, will trigger an expensive calculation.

However, the calculation will be triggered immediately. Would it make sense to allow to provide a default like this

    const [value] = useDebounce(fastValue, 200, defaultValue)

so that

    `value = defaultValue

until the debounce triggers the change?

**Comment by xnimorz** [comment]
Hello, @bebbi 

`useDebounce` works using the next algorithm:

1) When it's called for the first time, it will be initialized by the first value:
````
function A({text}) {
  const [ value ] = useDebounce(text, 500);
  return value; 
}

ReactDOM.render(<A text="Hello world"/>); // The tree is: "Hello world"
```

2) When component updates, `useDebounce` starts the debounced delay to change the value:
````
function A({text}) {
  const [ value ] = useDebounce(text, 500);
  return value; 
}

ReactDOM.render(<A text="Hello world"/>); // The tree is: "Hello world"
ReactDOM.render(<A text="Next value"/>); // The tree is: "Hello world", but after 500ms delay it will be changed to "Next value"
```

Also, this example could help you: https://codesandbox.io/s/rr40wnropq 
This example uses async operation when `useDebounce` changes the value. And this example sets up the default value, which is `United`.

To sum up, when `useDebounce` is called for the first time, it sets up the first argument as the initial value.  When it called the next times, it starts the debounce delay. 

I have created the example with classic `useDebounce` usage, and with `useDebouncedCallback` for custom logic: https://codesandbox.io/s/yj2796pzx1

**Comment by bebbi** [comment]
Hi @xnimorz, thanks, yes exactly that behaviour of immediately setting the first value is why I was thinking it could be described with `useThrottle`.

If we could provide a separate, e.g. falsy, default value, we could prevent one extra api call in your example in the situation where our value changes within the debounce duration after initial load.

But yes, `useDebouncedCallback` can achieve this and I'm using it for a replacement.

**Comment by xnimorz** [comment]
I close the issue. 
If you have any questions or suggestion, let me know :)
